examples/custom-libp2p/index.js

```diff
@@ -0,0 +1,121 @@
+'use strict'
+
+const Libp2p = require('libp2p')
+const IPFS = require('ipfs')
+const TCP = require('libp2p-tcp')
+const MulticastDNS = require('libp2p-mdns')
+const WebSocketStar = require('libp2p-websocket-star')
+const Bootstrap = require('libp2p-railing')
+const SPDY = require('libp2p-spdy')
+const KadDHT = require('libp2p-kad-dht')
+const MPLEX = require('libp2p-mplex')
+const SECIO = require('libp2p-secio')
+const assert = require('assert')
+
+/**
+ * This is the generator we will use to generate our fully customized libp2p node.
+ *
+ * @param {*} _ipfsNode The ipfs node. This houses the PeerInfo and PeerBook that modules may need
+ * @param {*} _ipfsConfig The config that is fetched from the ipfs-repo
+ * @returns {Libp2p} Our new libp2p node
+ */
+const libp2pGenerator = (_ipfsNode, _ipfsConfig) => {
+  // Set convenience variables to clearly showcase some of the useful things that are available
+  const peerInfo = _ipfsNode._peerInfo
+  const peerBook = _ipfsNode._peerBook
```

Comment:
hmm, `_peerInfo` and `_peerBook` presumably are pefixed with underscores because they're considered private APIs - I don't think we should encourage people to use them like this!

Would it be simpler and also still solve the issue (https://github.com/libp2p/js-libp2p/issues/222) to just allow passing `peerInfo` and `peerBook` to transports in the same way we do for peer discovery modules? https://github.com/libp2p/js-libp2p/blob/master/src/index.js#L185

I see the value in potentially being able to subclass `Libp2p` and also passing in the IPFS config, but if we don't have a use case for those two features I'd be tempted to go for a simpler approach for now to directly address @pgte's issue and which is also beneficial to people who use libp2p directly too.

Reply:
Ah, you might have to ignore that last comment (apologies it's late and I should be asleep already), I just looked at the code being used https://github.com/ipfs-shipyard/peer-star-app/blob/master/src/transport/ipfs.js#L26 and the issue is not being able to use the same instance in both `transport` _and_ `peerDiscovery`.

Let me re-review this tomorrow.

Reply:
Unless you think it's valuable to have these accessible outside of libp2p node creation context I'd probably have these passed separately to the factory function rather than exposing a new public API. I'd rather not have to support people randomly mutating the peer info/book.

Reply:
I'll update it to pass those into the generator.

Reply:
I've updated the generator to take a single object argument which includes the peerInfo, peerBook, ipfs node options and ipfs node config. The example has also been updated and I fixed the conflicts with master. Should be ready for another look over.